### PR TITLE
Bc generalize attach barcodes

### PR DESCRIPTION
**Pull request by benjamincarlin**
### Purpose
_Please explain the purpose of this PR and include links to any GitHub issues that it fixes:_

- No issue is linked to this PR.
---
### Changes
_Please list out what major changes were made in this PR to address the issue:_

- No changes.
---
### Review Instructions
_Please provide instructions about how should a reviewer test/verify the changes in this PR:_

- No instructions.

---
### PR Checklist
_Please ensure the following when opening a PR:_

- [ ] This PR added or updated tests.
- [x] This PR updated docstrings or documentation.
- [x] This PR applied Python style guidelines, specifically followed [NumPy style docstrings](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html) for this repo.
- [x] This PR considered generalizability beyond our own use case.

---
### Follow-up Discussions
_Please append follow-up discussions and issues during the review process below:_

- No follow-up discussions.

### CODE REVIEW

File: src/sctools/platform.py [hunk 1]
```diff
@@ -629,3 +629,162 @@ def attach_barcodes(cls, args=None):
...
+    molecule_barcode = None
+    sample_barcode = None
+
+    @classmethod
+    def _get_barcode(cls, barcode_start_pos, barcode_length, barcode_quality_tag, barcode_sequence_tag):
+        barcode = {"start": barcode_start_pos,
+                   "end": barcode_length,
+                   "quality_tag": barcode_quality_tag,
+                   "sequence_tag": barcode_sequence_tag}
+        return fastq.EmbeddedBarcode(**barcode)
```
Comment (danxmoran) [inline src/sctools/platform.py]: Using `**barcode` doesn't give any benefits here, would just pass the args directly:
```python
fastq.EmbeddedBarcode(start=barcode_start_pos, ...)
```

Comment (danxmoran) [reply src/sctools/platform.py]: In fact, do we even need this method? Could we replace all the spots where it's called with `fastq.EmbeddedBarcode` constructor calls?

File: src/sctools/platform.py [hunk 2]
```diff
@@ -629,3 +629,162 @@ def attach_barcodes(cls, args=None):
...
+
+        else:
+            if cls.cell_barcode and cls.molecule_barcode:
+                barcode_args["embedded_barcodes"] = [cls.cell_barcode,
+                                                     cls.molecule_barcode]
+            elif cls.cell_barcode:
+                barcode_args["embedded_barcodes"] = [cls.cell_barcode]
+            elif cls.molecule_barcode:
+                barcode_args["embedded_barcodes"] = [cls.molecule_barcode]
+            print(barcode_args)
```
Comment (danxmoran) [inline src/sctools/platform.py]: Remove this print?

File: src/sctools/platform.py [hunk 3]
```diff
@@ -629,3 +629,162 @@ def attach_barcodes(cls, args=None):
...
+                barcode_args["other_embedded_barcodes"] = cls.molecule_barcode
+
+        else:
+            if cls.cell_barcode and cls.molecule_barcode:
+                barcode_args["embedded_barcodes"] = [cls.cell_barcode,
+                                                     cls.molecule_barcode]
+            elif cls.cell_barcode:
+                barcode_args["embedded_barcodes"] = [cls.cell_barcode]
+            elif cls.molecule_barcode:
+                barcode_args["embedded_barcodes"] = [cls.molecule_barcode]
```
Comment (danxmoran) [inline src/sctools/platform.py]: These if-elif branches could be collapsed into a single:
```python
barcode_args["embedded_barcodes"] = [b for b in [cls.cell_barcode, cls.molecule_barcode] if b]
```

File: src/sctools/platform.py [hunk 4]
```diff
@@ -629,3 +629,162 @@ def attach_barcodes(cls, args=None):
...
+            tag_generators.append(fastq.EmbeddedBarcodeGenerator(**barcode_args))
+
+        return tag_generators
+
+    @classmethod
+    def attach_barcodes(cls, args=None):
+        parser = argparse.ArgumentParser()
+        parser.add_argument('--r1',
+                            required=True,
+                            help='read 1 fastq file for a 10x genomics v2 experiment')
```
Comment (danxmoran) [inline src/sctools/platform.py]: Remove "10x"?

File: src/sctools/platform.py [hunk 5]
```diff
@@ -629,3 +629,162 @@ def attach_barcodes(cls, args=None):
...
+                            help='filename for tagged bam')
+        parser.add_argument('-w',
+                            '--whitelist',
+                            default=None,
+                            help='optional cell barcode whitelist. If provided, corrected barcodes '
+                                 'will also be output when barcodes are observed within 1ED of a '
+                                 'whitelisted barcode')
+        parser.add_argument('--i1',
+                            default=None,
+                            help='(optional) i7 index fastq file for a 10x genomics experiment')
```
Comment (danxmoran) [inline src/sctools/platform.py]: Same, remove "10x"?

File: src/sctools/platform.py [hunk 6]
```diff
@@ -629,3 +629,162 @@ def attach_barcodes(cls, args=None):
...
+            args = parser.parse_args(args)
+        else:
+            args = parser.parse_args()
+
+        if ((bool(args.cell_barcode_start_pos) or args.cell_barcode_start_pos == 0)
+                != bool(args.cell_barcode_length) or
+            (bool(args.molecule_barcode_start_pos) or args.molecule_barcode_start_pos == 0)
+                != bool(args.molecule_barcode_length) or
+            (bool(args.sample_barcode_start_pos) or args.sample_barcode_start_pos == 0)
+                != bool(args.sample_barcode_length)):
```
Comment (danxmoran) [inline src/sctools/platform.py]: I'd extract the repeated bit of each clause (`(bool(thing) or thing == 0) != bool(other_thing)`) into a method, to make this less hairy

File: src/sctools/platform.py [hunk 7]
```diff
@@ -629,3 +629,162 @@ def attach_barcodes(cls, args=None):
         cls._tag_bamfile(args.u2, args.output_bamfile, tag_generators)
 
         return 0
+
+
+class Attach(GenericPlatform):
```
Comment (danxmoran) [inline src/sctools/platform.py]: Would name this something other than `Attach`... maybe `BarcodePlatform`?

File: setup.py [hunk 8]
```diff
@@ -37,6 +37,7 @@
     ],
     entry_points={
             'console_scripts': [
+                'AttachBarcodes = sctools.platform:BarcodePlatform.attach_barcodes',
```
Comment (samanehsan) [inline setup.py:40]: I think `TenXV2` can inherit from `BarcodePlatform` now since a lot of the logic looks the same

**Comment by samanehsan** [comment]
It looks like there are a couple of merge conflicts to resolve :octocat:

**Comment by samanehsan** [comment]
Do you have tests for this class? Or is it possible to modify the existing tests for attaching barcodes?

**Comment by samanehsan** [comment]
Can you also please fill out the PR template and link to a github issue if there is one? The PR description can be very brief, but I find it useful for understanding the context of the PR 🙂

**Pull request merged**